secrets/kubernetes/watch.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/log"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/prometheus/prometheus/secrets"
+)
+
+type WatchSPConfig struct {
+	ClientConfig
+}
+
+// Name returns the name of the Config.
+func (*WatchSPConfig) Name() string { return "kubernetes_watch" }
+
+// NewDiscoverer returns a Discoverer for the Config.
+func (c *WatchSPConfig) NewProvider(ctx context.Context, opts secrets.ProviderOptions) (secrets.Provider[SecretConfig], error) {
+	client, err := c.ClientConfig.client()
+	if err != nil {
+		return nil, err
+	}
+	return newWatchProvider(ctx, opts.Logger, client)
+}
+
+type watcher struct {
+	mu       sync.Mutex
+	w        watch.Interface
+	refCount uint
+	s        *corev1.Secret
+}
+
+func newWatcher(ctx context.Context, logger log.Logger, client kubernetes.Interface, config *SecretConfig) (*watcher, error) {
+	val := &watcher{
+		refCount: 1,
+	}
+	var err error
+	val.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(metav1.ObjectNameField, config.Name).String(),
+	})
+	if err != nil {
+		return val, fmt.Errorf("unable to watch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	val.s, err = client.CoreV1().Secrets(config.Namespace).Get(ctx, config.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return val, fmt.Errorf("unable to fetch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	go func() {
+		for {
+			select {
+			case e := <-val.w.ResultChan():
+				val.update(ctx, logger, client, e)
+			case <-ctx.Done():
+				val.w.Stop()
+				return
+			}
+		}
+	}()
+
+	return val, nil
+}
+
+func (w *watcher) update(ctx context.Context, logger log.Logger, client kubernetes.Interface, e watch.Event) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if e.Type == "" && e.Object == nil {
+		return
+	}
+	switch e.Type {
+	case watch.Modified, watch.Added:
+		secret := e.Object.(*corev1.Secret)
+		w.s = secret
+	case watch.Deleted:
+		w.s = nil
+	case watch.Bookmark:
```

Comment:
what this even means 🙈

Reply:
ah https://stackoverflow.com/questions/66080942/what-k8s-bookmark-solves

Reply:
Disabled this via a flag!